File: LeetCode/CPP/35.search-insert-position.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode id=35 lang=cpp
 *
 * [35] Search Insert Position
 */
#include <bits/stdc++.h>
#include <vector>
// @lc code=start
class Solution
{
public:
    int searchInsert(vector<int> &nums, int target)
    {
        long long l = 0, h = nums.size();
        long long m;
        while (l < h)
        {
            // To avoid Overflow
            m = l + (h - l) / 2;

            // Return index if element is found in the array
            if (nums[m] == target)
            {
                return m;
            }
            // Since the number is bigger than the target,
            else if (nums[m] > target)
            {
                h = m;
            }
            else
            {
                l = m + 1;
            }
        }
        return l;
    }
};
// @lc code=end

File: LeetCode/Rust/35.search-insert-position.rs
impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let mut lo: usize = 0;
        let mut hi: usize = nums.len() as usize;
        let mut mid: usize = 0;
        while lo < hi {
            mid = lo + (hi - lo) / 2;
            if nums[mid] == target {
                mid as i32;
            } else if nums[mid] > target {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        lo as i32
    }
}